File: src/Python/92.ReverseLinkedListII.py

class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution:
    def reverseBetween(self, head: ListNode, m: int, n: int) -> ListNode:
        if m == n:
            return head
        i = 1
        node = head
        beginnode = head
        tmp = None
        lastnode = None
        nextnode = None
        while node is not None:
            if i < m:
                lastnode = node
                nextnode = node.next
            elif i == m:
                beginnode = node
                nextnode = node.next
                node.next = tmp  
                tmp = node
            elif i < n:
                nextnode = node.next
                node.next = tmp  
                tmp = node
            elif i == n:
                beginnode.next = node.next
                node.next = tmp 
                if lastnode is None:
                    return node
                else:  
                    lastnode.next = node
                    return head
            i += 1
            node = nextnode
        return head

if __name__ == "__main__":
    solution = Solution()
    node = ListNode(1)
    node.next = ListNode(2)
    node.next.next = ListNode(3)
    node.next.next.next = ListNode(4)
    node.next.next.next.next = ListNode(5)
    node.next.next.next.next.next = ListNode(6)
    print(solution.reverseBetween(node, 2, 4))
   

File: src/Java/92.ReverseLinkedListII.java

import java.util.List;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashMap;
import java.util.Queue;
import java.util.HashSet;
import java.util.LinkedList;

public class ListNode {
    int val;
    ListNode next;

    ListNode(int x) {
        val = x;
    }
}

class Solution {
    public ListNode reverseBetween(ListNode head, int m, int n) {
        if (m == n) 
            return head;
        ListNode dum = new ListNode(0);
        dum.next = head;
        ListNode pre = dum;
        for (int i = 1; i < m; i++){
            pre = pre.next;  
        }
        head = pre.next;
        for (int i = m; i < n; i++){
            ListNode nex = head.next;
            head.next = nex.next;
            nex.next = pre.next;
            pre.next = nex;
        }
        return dum.next;
    }
}
